evm/src/main/java/org/hyperledger/besu/evm/operation/PushOperation.java

```diff
@@ -75,6 +76,11 @@ public static OperationResult staticOperation(
     } else {
       final int copyLength = Math.min(pushSize, code.length - pc - 1);
       push = Bytes.wrap(code, copyStart, copyLength);
+      // Right Pad the push with 0s up to pushSize if greater than the copyLength
+      final int rightPad = pushSize - copyLength;
+      if (rightPad != 0) {
+        push = push.shiftLeft(rightPad * 8, MutableBytes.create(pushSize));
```

Comment:
Do we not need a similar behavior if code.length <= copyStart?  or will the ROM module treat a single zero byte as equivalent to `n` zero bytes?

Reply:
It will be equivalent for us

Reply:
shiftLeft has awful performance for what needs to be done here, just moving bytes around.

There is no arbitrary-lengh equivelant, but use a construct siliar to Bytes32.leftPad() - https://github.com/tmio/tuweni/blob/aeeb06bc12d4f21a725c74ec86e48396b7f23581/bytes/src/main/java/org/apache/tuweni/bytes/Bytes32.java#L152-L154 - except make the mutableBytes pushSize length.

Reply:
This will definitely add something to the execution time. My preference would be to not change it at all since there's no consensus issue and PUSH is quite hot. We might think this is less frequent than it actually is because I can see code being compacted by compiler legitemately for storage purposes. We could choose to handle it like a kind of virtual pushN without changing the stack. Could we back this out at some point?
I.e. I see this as an implementation detail as we could have actually stack compaction for instance to optimise things.

Anyhow if we still want to proceed the best way would be to create a Bytes object with MutableBytes initialised to 0 and then use 'set' to add the value on it. This prevents creation of extra objects here

Reply:
Here is what we do in Linea's project when we want to rightPad, because Tuweni library doesn't have an easy left/right padding: https://github.com/Consensys/linea-tracer/blob/624f9cf971c4ae2a8089c41696fdec98f520a411/arithmetization/src/main/java/net/consensys/linea/zktracer/types/Utils.java#L48-L51

Would you be ok with something like this ? If yes, where should I write this `Bytes` library extension method ?

Reply:
> We might think this is less frequent than it actually is but I could see code being compacted by compiler legitemately for storage purposes.

Why would an optimized compiler end on an incomplete push ? The same code without the final incomplete push would be more optimized, unless the purpose is to burn 3 gas on the final instruction ^^

Reply:
The Tuweni Bytes package should just be brought back into Besu (it started there when it was called pantheon).  That way when we need to add something such as right pad we do it directly on the MutableBytes object, instead of copying implementations into random places.

It also allows for a tighter optimization loop when fixing performance issues.

Reply:
> Why would an optimized compiler end on an incomplete `PUSH` ? The same code without the final incomplete `PUSH` would be functionally equivalent and more optimized, unless the purpose is to burn 3 gas on the final instruction ^^ The point being that the value being pushed cannot be used.

An optimized compiler wouldn't, but a troll would.  This is the sorts of defense in depth that will pay silent dividends.

Reply:
"Obsessive attention to the spec" is what led @letypequividelespoubelles and myself to detect this deviation from the Yellow paper and ask for this change. Our arithmetization of PUSH instructions as part of Linea's zkevm implements the EYP faithfully (or attempts to, at least) and clashed with Besu's handling of incomplete PUSH's.

Reply:
> For example, the best API would be copy(byte[] data, int offset, int len) and in the impl we would use System.arrayCopy to do the copy since it has access to the array, and the hotspot compiler has an intrinsic for it.

This is how copyTo works currently in ArrayWrappingBytes

Reply:
>The Tuweni Bytes package should just be brought back into Besu (it started there when it was called pantheon). That way when we need to add something such as right pad we do it directly on the MutableBytes object, instead of copying implementations into random places.

indeed +1 on bringing tuweni into besu or at least have same maintainers as in besu so we can all make updates to it. Having code modularised is also not bad though, TBH. This framework sits at the core of the besu client, I don't understand how is it close to impossible to modify it. It's not even modernized anymore, still builds with java 8 or 11.

>For example, the best API would be copy(byte[] data, int offset, int len) and in the impl we would use System.arrayCopy to do the copy since it has access to the array, and the hotspot compiler has an intrinsic for it.

There's `MutableArrayWrappingBytes.set(int, Bytes)` which uses `System.arraycopy` but no `MutableArrayWrappingBytes.set(int, byte[])` or similar. I don't see how a `copy` method would help because in the end you have to return `Bytes` so it's another object to create when you return, unless you are referring to `MutableBytes`, but `set` would be a better name.

What I think we need to do is work a lot more with `MutableBytes` and less with `Bytes` - only work with `Bytes` if you are storing it as a final field in a class (i.e. it represents data and not state) - and we need to make some API changes in tuweni to motivate developers to do this.